### PR TITLE
Facet path string

### PR DESCRIPTION
**Pull request by halvorboe**
Fix for https://github.com/tantivy-search/tantivy/issues/580.

### CODE REVIEW

**Comment by fulmicoton** [comment]
@halvorboe your impl is incorrect, but this is probably due to confusing existing unit test.

The existing implementation in display is correct. Could you simply copy paste the logic from fmt::Display and call your method in fmt::Display. No need to care about the extra allocs.

Explanation:
There are only two representations of Facets:

The text one is //toto/titi.
The different step of the path may contain / or \ in which case it needs to be escaped by \

The "binary" one is 0u8 separated. The steps are not escaped.

File: src/schema/facet.rs [hunk 1]
```diff
@@ -122,6 +122,11 @@ impl Facet {
     pub fn to_path(&self) -> Vec<&str> {
         self.encoded_str().split(|c| c == FACET_SEP_CHAR).collect()
     }
+
+    /// Extract the path as a string from the `Facet`.
+    pub fn to_path_string(&self) -> String {
+        "//".to_owned() + &self.to_path().join("/")
```
Comment (fulmicoton) [inline src/schema/facet.rs]: Escaping is missing here. See comment in the Conversation tab of this PR.

File: src/schema/facet.rs [hunk 2]
```diff
@@ -122,6 +122,11 @@ impl Facet {
     pub fn to_path(&self) -> Vec<&str> {
         self.encoded_str().split(|c| c == FACET_SEP_CHAR).collect()
     }
+
+    /// Extract the path as a string from the `Facet`.
+    pub fn to_path_string(&self) -> String {
+        format!("/{}", self.to_string())
```
Comment (fulmicoton) [inline src/schema/facet.rs]: The initial double slash is not a bad idea, but this is not the way `Facet` have been designed.

We want `.to_path_string()` to be the opposite of `Facet::from(&str)`.

File: src/schema/facet.rs [hunk 3]
```diff
@@ -265,4 +270,21 @@ mod tests {
...
         assert_eq!(facet.to_path(), v);
     }
+
+    #[test]
+    fn test_to_path_string() {
+        let v = ["first", "second", "third/not_fourth"];
+        let facet = Facet::from_path(v.iter());
+        assert_eq!(
+            facet.to_path_string(),
+            String::from("//first/second/third\\/not_fourth")
```
Comment (fulmicoton) [inline src/schema/facet.rs]: `/first/second/third\\/not_fourth`

File: src/schema/facet.rs [hunk 4]
```diff
@@ -265,4 +270,21 @@ mod tests {
...
+        );
+    }
+
+    #[test]
+    fn test_to_path_string_empty() {
+        let v: Vec<&str> = vec![];
+        let facet = Facet::from_path(v.iter());
+        assert_eq!(facet.to_path_string(), "//");
```
Comment (fulmicoton) [inline src/schema/facet.rs]: `/`

File: src/schema/facet.rs [hunk 5]
```diff
@@ -122,6 +122,11 @@ impl Facet {
     pub fn to_path(&self) -> Vec<&str> {
         self.encoded_str().split(|c| c == FACET_SEP_CHAR).collect()
     }
+
+    /// Extract the path as a string from the `Facet`.
```
Comment (fulmicoton) [inline src/schema/facet.rs]: /// This function is the inverse of `Facet::from(&str)`.

File: src/schema/facet.rs (hunk 2 above)
Comment (halvorboe) [reply src/schema/facet.rs]: My bad, was due to the first comment you made on this issue. Can see what the original issue said now.

**Pull request merged**